### PR TITLE
[Tech] Let's be even more strict when writing TS code

### PR DESCRIPTION
**Pull request by CommandMC**
I've changed some of our TS settings so we're forced to account for more situations that were easy to miss before. I think this will ultimately lead to a more stable experience for our users, at the cost of having to do *a bit* more work when writing new code

Changes in tsconfig.json:
- Ordered options based on where they appear in TS' docs
- Added the following options:
   - `allowUnreachableCode`: `false`
   - `allowUnusedLabels`: `false`
   - `noUncheckedIndexedAccess`: `true`
- Turned off the following options:
   - `suppressImplicitAnyIndexErrors`
- Options that were removed but whose values were set to their default values:
  (Just to clarify, these changes only cleaned up our config file, no functional changes were made)
   - `noImplicitThis`: Already `true` since `strict` is enabled
   - `noImplicitAny`: Same as above
   - `strictNullChecks`: Same as above

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [X] Tested the feature and it's working on a current and clean install.
- [ ] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

### CODE REVIEW

File: src/backend/utils.ts [hunk 1]
```diff
@@ -365,7 +365,8 @@ async function errorHandler(
   const otherErrorMessages = ['No saved credentials', 'No credentials']
 
   if (!window) {
-    window = BrowserWindow.getFocusedWindow() ?? BrowserWindow.getAllWindows[0]
+    window =
+      BrowserWindow.getFocusedWindow() ?? BrowserWindow.getAllWindows()[0]
```
Comment (CommandMC) [inline src/backend/utils.ts:369]: As an example on why these changes are necessary, take a look at this diff. Before, we were trying to access the first element of `BrowserWindow.getAllWindows`, which would of course never have worked since it's a function and not an array, but we weren't warned about that since `suppressImplicitAnyIndexErrors` was on

File: src/backend/gog/games.ts [hunk 2]
```diff
@@ -159,7 +159,14 @@ class GOGGame extends Game {
     const etaMatch = data.match(/ETA: (\d\d:\d\d:\d\d)/m)
     const bytesMatch = data.match(/Downloaded: (\S+) MiB/m)
     const progressMatch = data.match(/Progress: (\d+\.\d+) /m)
-    if (etaMatch && bytesMatch && progressMatch) {
+    if (
+      etaMatch &&
```
Comment (Nocccer) [inline src/backend/gog/games.ts:163]: If the array gets longer you could do:
```typescript
!etaMatch.some((entry) => entry === null))
```

File: src/backend/utils.ts [hunk 3]
```diff
@@ -95,33 +95,33 @@ function semverGt(target: string, base: string) {
...
   const [tmajor, tminor, tpatch] = target.split('.').map(Number)
 
   let isGE = false
   // A pretty nice piece of logic if you ask me. :P
-  isGE ||= tmajor > bmajor
-  isGE ||= tmajor === bmajor && tminor > bminor
-  isGE ||= tmajor === bmajor && tminor === bminor && tpatch > bpatch
+  isGE ||= tmajor! > bmajor!
+  isGE ||= tmajor === bmajor && tminor! > bminor!
+  isGE ||= tmajor === bmajor && tminor === bminor && tpatch! > bpatch!
```
Comment (Nocccer) [inline src/backend/utils.ts:124]: I am not sure that this can not be undefined. We actually wrote a test for it where both passed values are undefined. I somehow don't like the `!` operator. With this you trust your code to much. Better are checks and tests.

**Review by Nocccer — CHANGES_REQUESTED** [review]
Looks good overall. 
Still have some questions/suggestions.

File: src/backend/gog/games.ts (hunk 2 above)
Comment (CommandMC) [reply src/backend/gog/games.ts:163]: Not sure if that would help here, since I have to check that
1. `etaMatch` itself isn't `null`
2. the group provided in the RegExp (`etaMatch[1]`) matched something
I believe both of those cases can happen independently, at least TS warned me about it

File: src/backend/utils.ts (hunk 3 above)
Comment (CommandMC) [reply src/backend/utils.ts:124]: I'd argue that it would be the responsibility of the one calling this function to check the inputs first (in this case, making sure you're actually passing SemVer strings). But I agree, a simple check isn't too much to ask

**Comment by CommandMC** [comment]
Arg, got caught by polymorphism here... `0` is of course also `false`. Now what's the most elegant way to check just for `undefined`...

File: src/backend/gog/games.ts (hunk 2 above)
Comment (Nocccer) [reply src/backend/gog/games.ts:163]: Should you then not test `etaMatch && etaMatch[0] && etaMatch[1]`. I don't know if typescript arrays work like C arrays where `etaMatch` would be the pointer to `etaMatch[0]`

File: src/backend/utils.ts (hunk 3 above)
Comment (Nocccer) [reply src/backend/utils.ts]: I would not take responsibility to the user. We working in a open source project where everyone have his design choices. I try to catch all possible errors, to avoid any traces and actually that should be our goal, else we never get stable. We got now 3 renderer traces causing empty screen in frontend.

Btw what happens if you still use the operator and write a test with both versions undefined. Does it return false?

**Comment by Nocccer** [comment]
> Arg, got caught by polymorphism here... `0` is of course also `false`. Now what's the most elegant way to check just for `undefined`...

Someone in the internet said checking for null also catches undefined.

**Comment by CommandMC** [comment]
I ended up just using the `semver` library, I think some dependency we have already requires it anyways and it's far easier to just use it rather than implementing everything ourselves

File: src/backend/gog/games.ts (hunk 2 above)
Comment (CommandMC) [reply src/backend/gog/games.ts:163]: As long as `null` isn't returned, `etaMatch[0]` will always be the matched text. See [the spec on `RegExp.exec`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#return_value) (called by `String.match` internally)

Comment (Nocccer) [reply src/backend/gog/games.ts:163]: Ok yes makes sense. This is fine then, aslong as we don't do something like this:
```typescript
if(etaMatch && etaMatch[1] && etaMatch[2] && etaMatch[3] && ...)
```

File: src/backend/gog/games.ts [hunk 4]
```diff
@@ -159,15 +159,8 @@ class GOGGame extends Game {
...
-      etaMatch &&
-      etaMatch[1] &&
-      bytesMatch &&
-      bytesMatch[1] &&
-      progressMatch &&
-      progressMatch[1]
-    ) {
-      const eta = etaMatch[1]
+    if (bytesMatch && bytesMatch[1] && progressMatch && progressMatch[1]) {
+      const eta = etaMatch ? etaMatch[1] : null
```
Comment (Nocccer) [inline src/backend/gog/games.ts:163]: If we using only the index one, is it not enough to check for those?

Comment (CommandMC) [reply src/backend/gog/games.ts:163]: Just checking `bytesMatch[1]` for example?  
No, that would not be enough. `bytesMatch` itself can be null, and trying to access properties of null will throw a TypeError

Comment (arielj) [reply src/backend/gog/games.ts:163]: We can do `if (bytesMatch?.[1] && progressMatch?.[1])` which looks kinda ugly but it actually works xD

File: src/frontend/state/GlobalState.tsx [hunk 5]
```diff
@@ -92,9 +92,9 @@ export class GlobalState extends PureComponent<Props> {
       []
     for (const igame in games) {
       for (const installedGame of installedGames) {
-        if (installedGame.appName === games[igame].app_name) {
-          games[igame].install = installedGame
-          games[igame].is_installed = true
+        if (installedGame.appName === games[igame]!.app_name) {
```
Comment (flavioislima) [inline src/frontend/state/GlobalState.tsx:95]: should not be an interrogation mark instead of a exclamation one to see if the item is undefined?

File: src/frontend/screens/WebView/index.tsx [hunk 6]
```diff
@@ -32,27 +32,17 @@ export default function WebView() {
...
-    '/loginweb/legendary': epicLoginUrl,
-    '/loginweb/gog': gogLoginUrl
+  const urls: { [index: string]: string } = {
+    '/epicstore': `https://www.epicgames.com/store/${lang}/`,
+    '/gogstore': `https://gog.com`,
+    '/wiki':
+      'https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/wiki',
+    '/loginweb/legendary': 'https://legendary.gl/epiclogin',
+    '/loginweb/gog':
+      'https://auth.gog.com/auth?client_id=46899977096215655&redirect_uri=https%3A%2F%2Fembed.gog.com%2Fon_login_success%3Forigin%3Dclient&response_type=code&layout=galaxy'
```
Comment (flavioislima) [inline src/frontend/screens/WebView/index.tsx:45]: Don't you think this made it more unreadable?
Another idea would be to move the list of URLs to a constants file and import them, making the file more cleaner.

File: src/frontend/screens/Settings/components/Tools/index.tsx [hunk 7]
```diff
@@ -70,14 +70,15 @@ export default function Tools() {
...
-      if (ev.dataTransfer.items[0].kind === 'file') {
-        const exe = ev.dataTransfer.items[0].getAsFile()?.path
-        if (exe) {
-          return callTools('runExe', exe)
-        }
+    if (!ev.dataTransfer.items[0]) {
+      return
+    }
+    // Use DataTransferItemList interface to access the file(s)
+    // If dropped items aren't files, reject them
```
Comment (flavioislima) [inline src/frontend/screens/Settings/components/Tools/index.tsx:77]: just a reminder here to remove this Drag functionality. I don't think it is pretty useful. Maybe let's just add a new button like the others and select the EXE on click.

File: src/frontend/screens/Library/components/InstallModal/index.tsx [hunk 8]
```diff
@@ -81,7 +81,7 @@ function getInstallLanguage(
       return foundAvailable
     }
   }
-  return availableLanguages[0]
+  return availableLanguages[0]!
```
Comment (flavioislima) [inline src/frontend/screens/Library/components/InstallModal/index.tsx:84]: What is the name of this feature of adding an exclamation mark at the end? I don't remember using this one before.

File: src/frontend/screens/Game/GameSubMenu/index.tsx [hunk 9]
```diff
@@ -153,7 +153,7 @@ export default function GamesSubmenu({
         defaultPath: defaultInstallPath
       })
       if (path) {
-        await window.api.changeInstallPath([appName, path, runner])
+        await window.api.changeInstallPath(appName, runner, path)
```
Comment (flavioislima) [inline src/frontend/screens/Game/GameSubMenu/index.tsx:156]: If changing it from an array maybe using an object would be better since we won't need to care about the order of the arguments.

File: src/backend/utils.ts [hunk 10]
```diff
@@ -54,46 +55,10 @@ const { showMessageBox } = dialog
...
-  const [bmajor, bminor, bpatch] = base.split('.').map(Number)
-  const [tmajor, tminor, tpatch] = target.split('.').map(Number)
-
-  let isGE = false
-  // A pretty nice piece of logic if you ask me. :P
-  isGE ||= tmajor > bmajor
-  isGE ||= tmajor === bmajor && tminor > bminor
-  isGE ||= tmajor === bmajor && tminor === bminor && tpatch > bpatch
-  return isGE
+  return semver.gt(target, base)
```
Comment (flavioislima) [inline src/backend/utils.ts:61]: we need to test if the package covers all use cases. I am worried about when we release a stable from a previous beta.
like:
- 2.2.0-beta.1, then
- 2.2.0

**Review by flavioislima — COMMENTED** [review]
It is a lot of changes, will test it during the week.
Left a few comments for now.

**Review by redromnon — COMMENTED** [review]
LGTM :+1:

File: src/backend/utils.ts (hunk 10 above)
Comment (CommandMC) [reply src/backend/utils.ts:61]: There are tests for this function, testing exactly that. They're all passing

File: src/frontend/screens/Library/components/InstallModal/index.tsx (hunk 8 above)
Comment (CommandMC) [reply src/frontend/screens/Library/components/InstallModal/index.tsx:84]: It's the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator). It's basically a way of telling Typescript "this thing can't be undefined/null right now", in the rare case that it's not able to detect this on its own

File: src/frontend/state/GlobalState.tsx (hunk 5 above)
Comment (CommandMC) [reply src/frontend/state/GlobalState.tsx:95]: We're looping over all indexes in `games` (see the `for (const igame in games)` above). Thus, it's not possible for `games[igame]` to be undefined here

File: src/frontend/screens/WebView/index.tsx (hunk 6 above)
Comment (CommandMC) [reply src/frontend/screens/WebView/index.tsx:45]: I mean, having the URLs as their own variables up top isn't very readable either IMO  
Having a separate file would probably be a good idea, I agree

File: src/frontend/screens/Library/components/InstallModal/index.tsx (hunk 8 above)
Comment (flavioislima) [reply src/frontend/screens/Library/components/InstallModal/index.tsx:84]: Yes, true. I don't really use it that much because it has some issues imo:
- Adding this operator moves the responsibility to know if the prop can be null/undefined from TS to the developer.
- With that, we might actually cause something to fail because some argument or prop come null
- I like to deal with this like this is using `if` or optional chaining (not the most beautiful way but the safest imo:
`if(prop){
return prop
}`

or 

`if(something && something.value)` or `if(something?.value){`

The thing is that on this refactor there are dozens of cases where we needed this operator so might be a sign that we should maybe review some of our types or add more checks.

Comment (flavioislima) [reply src/frontend/screens/Library/components/InstallModal/index.tsx:84]: but I am not arguing against using it, it is actually pretty useful, we just need to be extra careful :P

Comment (CommandMC) [reply src/frontend/screens/Library/components/InstallModal/index.tsx:84]: I definitely agree that it should be the last thing you use, but there are simply cases where you *have* to use it since an undefined/null simply *can't* happen, and adding checks everywhere for situations that can't happen is pointless IMO

As a simple example, we will always have a BrowserWindow open, but `BrowserWindow.getAllWindows()` returns an array that can technically be empty. In cases like these, where it's *absolutely certain* that the given value isn't undefined/null, it's fine to use. There are just simply cases in a big project like this where the developer *can* know better than TypeScript, and the operator is meant for those situations

File: src/frontend/screens/Game/GameSubMenu/index.tsx (hunk 9 above)
Comment (CommandMC) [reply src/frontend/screens/Game/GameSubMenu/index.tsx:156]: I think we have a mix of both styles (regular arguments, and a single object containing everything) in the project right now, if we want to unify everything that's fine, but I'd say that belongs into a different PR

File: src/frontend/screens/WebView/index.tsx (hunk 6 above)
Comment (CommandMC) [reply src/frontend/screens/WebView/index.tsx]: I've ended up not putting them in a separate file, see notes on the commit for details

**Review by arielj — APPROVED** [review]
looks good to me!

**Comment by CommandMC** [comment]
Currently waiting on #1887 and #1901 to merge before updating this

**Comment by Poopooracoocoo** [comment]
Wvewvevvwv wWvewvevvwv w

**Comment by CommandMC** [comment]
I doubt I'm going to work on it anytime soon, might as well open up a new PR then
